### ISSUE TITLE
Add elsifdef and elsedef conditional directives

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#514]
AsciiDoc only supports ifdef and ifndef conditional directives. When you have several alternatives in your content, the lack of "else if" and "else" directives leads to very verbose conditional logic. Add the conditional directives `elsifdef` and `elsedef` to simplify this type of logic.

Consider the following:

```
ifdef::linux[]
Content for Linux users.
endif::linux[]

ifdef::windows[]
Content for Windows users.
endif::windows[]

ifdef::osx[]
Content for OSX users.
ifdef::osx[]

ifndef::linux,windows,osx[]
Content when no OS is specified.
endif::linux,windows,osx[]
```

This markup could be reduced drastically with the introduction of `elsifdef` and `else` conditional directives:

```
ifdef::linux[]
Content for Linux users.
elsifdef::windows[]
Content for Windows users.
elsifdef::osx[]
Content for OSX users.
elsedef::[]
Content when no OS is specified.
endif::[]
```

Open question. I'm not sure if the "else" directive should be `else::[]`, `elsedef::[]` or `elsifdef::[]` (no target).

I also think that `endifdef::[]` should be accepted in addition to `endif::[]`

**Comment by mojavelinux** [comment]
Here's the documentation of conditionals in the C preprocessor for reference:

https://gcc.gnu.org/onlinedocs/cpp/Conditionals.html

**Comment by vivien** [comment]
I think an _else_ preprocessor directive similar to cpp is the intuitive and KISS way to do it. The important thing though is to make sure it becomes part of the AsciiDoc syntax and not an Asciidoctor extension...

**Comment by oddhack** [comment]
Having just dealt with some nested ifdef conditionals, I'd find both an 'else' directive and the ability to say defined(attribute-name) or equivalent in ifeval::, very useful.

**Comment by mojavelinux** [comment]
> The important thing though is to make sure it becomes part of the AsciiDoc syntax and not an Asciidoctor extension...

If it's added to Asciidoctor, it's part of the AsciiDoc language.

**Comment by marcdexet-cnrs** [comment]
+1

**Comment by bcouetil** [comment]
I would love to see this too, whatever the keywords.

Any progress on this ?

**Comment by mojavelinux** [comment]
This proposal is being considered for the AsciiDoc Language project (which is finallyl getting underway). That's where it has to start.

**Comment by pobanion** [comment]
Hi...  I was just wondering if there are any updates on having "elseif...else" in asciidocs?

**Comment by matthieu-labas** [comment]
I'd +1 that too. I use `ifeval` constructs to handle translations in the same file and it's very copy-paste-error prone:

```
ifeval::["{lang}" == "fr"]
Bonjour
endif:[]
ifeval::["{lang}" != "fr"] # see the potential for copy-paste error here?
Hello
endif:[]
```

It would be easier and more powerful to be able to write:

```
ifeval::["{lang}" == "fr"]
Bonjour
elsifeval::["{lang}" == "es"]
Hola
elseval:[] # English by default
Hello
endif:[]
```

Unless there are other constructs that allow for this?

**Comment by calculuswhiz** [comment]
@matthieu-labas In your case (translation files), but you may have success setting up like this (safe mode off).

## Main layout

Combine the `include` with attributes as noted [here](https://docs.asciidoctor.org/asciidoc/latest/directives/include/)

**Layout.adoc**
```adoc
// Select language attribute files
include::{lang}.adoc
= {Chapter} 1

{greeting}
```

## Attribute files

**ga.adoc**
```adoc
:Chapter: Caibidil
:greeting: Dia duit
```

**en.adoc**
```adoc
:Chapter: Chapter
:greeting: Hello
```

## Alternatives

Some alternatives to `include`, if you're OK with preprocessing the files via scripts:
- Concatenate the file for the current language with the layout file then process that via the API.
- Or keep everything in the same file, keeping attribute translations distinguished by suffix. Preprocess the file eliminating the suffix of your current output language. Then concatenate this with the layout file and process it:
  ```adoc
  // Attributes file to preprocess
  :Chapter_ga: Caibidil
  :Chapter_en: Chapter
  :greeting_ga: Dia duit
  :greeting_en: Hello
  ```

Yes, I ran into this too. For a couple languages, the last one works fine, but I don't think it scales very well with more languages and collaborators, so I will try to switch the individual files when I add more languages.

**Comment by matthieu-labas** [comment]
@calculuswhiz the idea was for me to have the structure (chapters, paragraphs) common to all languages, then each paragraph has all languages one after another. Having a separate file for each language means I can easily forget to add a new paragraph in another language; if I have everything at the same place I add the translations at the same time I add a new paragraph.

Though of course it will become messy if more languages are added. In my case I only handle two translation: my native language and english, so it's manageable, but a `elseval:[]` would be nice...

**Comment by calculuswhiz** [comment]
@matthieu-labas  To be clear, I totally agree. Else branching seems like a clear advantage, and I was very surprised when I learned it didn't exist (which led me here). The macros and meta-programming aspects are underwhelming, and I have had to do too much external programming to get things to work.

Lately, I've been brainstorming ideas of generating .adoc attribute files with XLIFF or other translation file formats and managing the source with an editor. I'm still researching this though.

**Comment by tovine** [comment]
I agree with the above. Found this issue after wondering where `else::[]` was.
How has this been open for over 13 years with no resolution yet?
Please just go with `else` and `endif` as that should catch a lot of useful cases, and then `elsif/def/eval` could follow later as an enhancement if needed.
